File: frontend/bff/src/infrastructure/grpc/mappers/mandate.contract-mapper.ts
import type { MandateTermsInput } from '../../../application/mandates/commands/mandate-terms.input.js';
import type { ComplianceDto } from '../../../application/mandates/dtos/compliance.dto.js';
import type { MandateDto } from '../../../application/mandates/dtos/mandate.dto.js';
import type { MandateTermsDto } from '../../../application/mandates/dtos/mandate-terms.dto.js';
import { nullable, nullableNumber } from './common.contract-mapper.js';
import { commodityEnum, complianceStatusEnum, mandateStatusEnum, mandateTypeEnum, measurementUnitEnum } from './enum.contract-mapper.js';

export interface ContractCompliance {
  status: string;
  reason: string;
}

export interface ContractMandateTerms {
  title: string;
  criteria?: string;
  commodity: string;
  tenor?: string;
  quantity?: number;
  quantityUnit: string;
  price: { atMarket: boolean; target?: number; min?: number; max?: number; unit?: string } | null;
  windowStart?: string;
  windowEnd?: string;
}

export interface ContractMandate {
  id: string;
  policyId: string;
  policyCode: string;
  policyVersion: string;
  axisId: string;
  axisCode: string;
  axisTitle: string;
  type: string;
  terms: ContractMandateTerms;
  consumed: number;
  balance?: number;
  compliance: ContractCompliance;
  status: string;
  issuedBy: string;
  decidedBy?: string;
  decidedAt?: string;
  decisionNote?: string;
}

export const toComplianceDto = (c: ContractCompliance): ComplianceDto => ({
  status: complianceStatusEnum.fromContractRequired(c.status),
  reason: c.reason,
});

const toTermsDto = (t: ContractMandateTerms): MandateTermsDto => ({
  title: t.title,
  criteria: nullable(t.criteria),
  commodity: commodityEnum.fromContract(t.commodity),
  tenor: nullable(t.tenor),
  quantity: nullableNumber(t.quantity),
  quantityUnit: measurementUnitEnum.fromContractRequired(t.quantityUnit),
  price: {
    atMarket: t.price?.atMarket ?? false,
    target: nullableNumber(t.price?.target),
    min: nullableNumber(t.price?.min),
    max: nullableNumber(t.price?.max),
    unit: nullable(t.price?.unit),
  },
  windowStart: nullable(t.windowStart),
  windowEnd: nullable(t.windowEnd),
});

export const toMandateDto = (m: ContractMandate): MandateDto => ({
  id: m.id,
  policyId: m.policyId,
  policyCode: m.policyCode,
  policyVersion: m.policyVersion,
  axisId: m.axisId,
  axisCode: m.axisCode,
  axisTitle: m.axisTitle,
  type: mandateTypeEnum.fromContractRequired(m.type),
  terms: toTermsDto(m.terms),
  consumed: m.consumed,
  balance: nullableNumber(m.balance),
  compliance: toComplianceDto(m.compliance),
  status: mandateStatusEnum.fromContractRequired(m.status),
  issuedBy: m.issuedBy,
  decidedBy: nullable(m.decidedBy),
  decidedAt: nullable(m.decidedAt),
  decisionNote: nullable(m.decisionNote),
});

export const toContractMandateTerms = (t: MandateTermsInput) => ({
  ...t,
  commodity: commodityEnum.toContract(t.commodity),
  quantityUnit: measurementUnitEnum.toContract(t.quantityUnit),
  price: { ...t.price },
});

File: backend/core-service/src/Fix.Presentation/Mappers/MandateContractMapper.cs
using Fix.Application.Mandates.Commands;
using Fix.Application.Mandates.Dtos;
using Contract = Fix.Contracts.V1;
using DomainCommon = Fix.Domain.Common;

namespace Fix.Presentation.Mappers;

internal static class MandateContractMapper
{
    public static Contract.Mandate ToContract(this MandateDto mandate)
    {
        var contract = new Contract.Mandate
        {
            Id = mandate.Id.ToString(),
            Code = mandate.Code,
            PolicyId = mandate.PolicyId.ToString(),
            PolicyCode = mandate.PolicyCode,
            PolicyVersion = mandate.PolicyVersion,
            AxisId = mandate.AxisId.ToString(),
            AxisCode = mandate.AxisCode,
            AxisTitle = mandate.AxisTitle,
            Type = mandate.Type.ToContract<Contract.MandateType>(),
            Terms = mandate.ToContractTerms(),
            Consumed = (double)mandate.Consumed,
            Compliance = mandate.Compliance.ToContract(),
            Status = mandate.Status.ToContract<Contract.MandateStatus>(),
            IssuedBy = mandate.IssuedBy.ToString(),
        };

        if (mandate.Balance is { } balance) contract.Balance = (double)balance;
        if (mandate.DecidedBy is { } decidedBy) contract.DecidedBy = decidedBy.ToString();
        if (mandate.DecidedAt is { } decidedAt) contract.DecidedAt = decidedAt.ToContract();
        if (mandate.DecisionNote is { } note) contract.DecisionNote = note;
        return contract;
    }

    public static Contract.Compliance ToContract(this ComplianceDto compliance) => new()
    {
        Status = compliance.Status.ToContract<Contract.ComplianceStatus>(),
        Reason = compliance.Reason,
    };

    public static MandateTermsInput ToInput(this Contract.MandateTerms? terms)
    {
        terms ??= new Contract.MandateTerms();
        var price = terms.Price ?? new Contract.PriceCriteria();

        return new MandateTermsInput(
            terms.Title,
            terms.Criteria.ToOptionalString(terms.HasCriteria),
            terms.Commodity.ToOptionalDomain<DomainCommon.Commodity>("terms.commodity"),
            terms.Tenor.ToOptionalString(terms.HasTenor),
            terms.Quantity.ToOptionalDecimal(terms.HasQuantity),
            terms.QuantityUnit.ToOptionalDomain<DomainCommon.MeasurementUnit>("terms.quantity_unit"),
            price.AtMarket,
            price.Target.ToOptionalDecimal(price.HasTarget),
            price.Min.ToOptionalDecimal(price.HasMin),
            price.Max.ToOptionalDecimal(price.HasMax),
            price.Unit.ToOptionalString(price.HasUnit),
            terms.WindowStart.ToOptionalString(terms.HasWindowStart).ToOptionalDate("terms.window_start"),
            terms.WindowEnd.ToOptionalString(terms.HasWindowEnd).ToOptionalDate("terms.window_end"));
    }

    private static Contract.MandateTerms ToContractTerms(this MandateDto mandate)
    {
        var price = new Contract.PriceCriteria { AtMarket = mandate.Price.AtMarket };
        if (mandate.Price.Target is { } target) price.Target = (double)target;
        if (mandate.Price.Min is { } min) price.Min = (double)min;
        if (mandate.Price.Max is { } max) price.Max = (double)max;
        if (mandate.Price.Unit is { } unit) price.Unit = unit;

        var terms = new Contract.MandateTerms
        {
            Title = mandate.Title,
            Commodity = mandate.Commodity.ToContract<Contract.Commodity>(),
            QuantityUnit = mandate.QuantityUnit.ToContract<Contract.MeasurementUnit>(),
            Price = price,
        };

        if (mandate.Criteria is { } criteria) terms.Criteria = criteria;
        if (mandate.Tenor is { } tenor) terms.Tenor = tenor;
        if (mandate.Quantity is { } quantity) terms.Quantity = (double)quantity;
        if (mandate.WindowStart is { } start) terms.WindowStart = start.ToContract();
        if (mandate.WindowEnd is { } end) terms.WindowEnd = end.ToContract();
        return terms;
    }
}
